Lets start by loading our dataset from KAGGLE.

In [ ]:
pip install kagglehub

In [ ]:
import kagglehub
import pandas as pd

# 2. Download the latest version of the dataset
path = kagglehub.dataset_download("subashmaster0411/social-media-engagement-dataset")

print("Path to dataset files:", path)

100%|██████████| 1.30M/1.30M [00:00<00:00, 46.9MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/subashmaster0411/social-media-engagement-dataset/versions/1


In [ ]:
import os

# List all files in the downloaded directory to find the exact CSV filename
files = os.listdir(path)
print("Files in directory:", files)

Files in directory: ['Social Media Engagement Dataset.csv']


In [ ]:
csv_file_name = files[0]
full_csv_path = os.path.join(path, csv_file_name)
df = pd.read_csv(full_csv_path)
df.head()

,post_id,timestamp,day_of_week,platform,user_id,location,language,text_content,hashtags,mentions,...,comments_count,impressions,engagement_rate,brand_name,product_name,campaign_name,campaign_phase,user_past_sentiment_avg,user_engagement_growth,buzz_change_rate
0,kcqbs6hxybia,2024-12-09 11:26:15,Monday,Instagram,user_52nwb0a6,"Melbourne, Australia",pt,Just tried the Chromebook from Google. Best pu...,#Food,NaN,...,701,18991,0.19319,Google,Chromebook,BlackFriday,Launch,0.0953,-0.3672,19.1
1,vkmervg4ioos,2024-07-28 19:59:26,Sunday,Twitter,user_ucryct98,"Tokyo, Japan",ru,Just saw an ad for Microsoft Surface Laptop du...,"#MustHave, #Food","@CustomerService, @BrandCEO",...,359,52764,0.05086,Microsoft,Surface Laptop,PowerRelease,Post-Launch,0.1369,-0.4510,-42.6
2,memhx4o1x6yu,2024-11-23 14:00:12,Saturday,Reddit,user_7rrev126,"Beijing, China",ru,What's your opinion about Nike's Epic React? ...,"#Promo, #Food, #Trending",NaN,...,643,8887,0.45425,Nike,Epic React,BlackFriday,Post-Launch,0.2855,-0.4112,17.4
3,bhyo6piijqt9,2024-09-16 04:35:25,Monday,YouTube,user_4mxuq0ax,"Lagos, Nigeria",en,Bummed out with my new Diet Pepsi from Pepsi! ...,"#Reviews, #Sustainable","@StyleGuide, @BrandSupport",...,743,6696,0.42293,Pepsi,Diet Pepsi,LaunchWave,Launch,-0.2094,-0.0167,-5.5
4,c9dkiomowakt,2024-09-05 21:03:01,Thursday,Twitter,user_l1vpox2k,"Berlin, Germany",hi,Just tried the Corolla from Toyota. Absolutely...,"#Health, #Travel","@BrandSupport, @InfluencerName",...,703,47315,0.08773,Toyota,Corolla,LocalTouchpoints,Launch,0.6867,0.0807,38.8


In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 28 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   post_id                  12000 non-null  object 
 1   timestamp                12000 non-null  object 
 2   day_of_week              12000 non-null  object 
 3   platform                 12000 non-null  object 
 4   user_id                  12000 non-null  object 
 5   location                 12000 non-null  object 
 6   language                 12000 non-null  object 
 7   text_content             12000 non-null  object 
 8   hashtags                 12000 non-null  object 
 9   mentions                 8059 non-null   object 
 10  keywords                 12000 non-null  object 
 11  topic_category           12000 non-null  object 
 12  sentiment_score          12000 non-null  float64
 13  sentiment_label          12000 non-null  object 
 14  emotion_type          

In [ ]:
print(df.describe())
print("---------------------------------------")
print("\nNULL VALUES SUM COLUMN WISE: \n\n",df.isnull().sum())

       sentiment_score  toxicity_score  likes_count  shares_count  \
count     12000.000000    12000.000000  12000.00000  12000.000000   
mean          0.000553        0.503868   2490.72025   1007.167167   
std           0.583563        0.288198   1441.53253    575.072282   
min          -0.999800        0.000000      0.00000      0.000000   
25%          -0.503200        0.251400   1236.00000    510.000000   
50%          -0.006200        0.505950   2496.00000   1018.000000   
75%           0.513525        0.756200   3723.25000   1501.000000   
max           0.999900        0.999900   5000.00000   2000.000000   

       comments_count   impressions  engagement_rate  user_past_sentiment_avg  \
count     12000.00000  12000.000000     12000.000000             12000.000000   
mean        504.34575  49811.338500         0.278137                 0.001472   
std         288.68416  28930.289451         1.149206                 0.576627   
min           0.00000    130.000000         0.001880  

In [ ]:
# Unload the interactive table just for this summary view to keep it clean
%unload_ext google.colab.data_table

# Format to 4 decimal places and add a soft blue heat map to highlight magnitude
summary_table = df.describe().style\
    .format("{:.4f}")\
    .background_gradient(cmap='Blues', axis=0)

summary_table

The google.colab.data_table extension is not loaded.


,sentiment_score,toxicity_score,likes_count,shares_count,comments_count,impressions,engagement_rate,user_past_sentiment_avg,user_engagement_growth,buzz_change_rate
count,12000.0000,12000.0000,12000.0000,12000.0000,12000.0000,12000.0000,12000.0000,12000.0000,12000.0000,12000.0000
mean,0.0006,0.5039,2490.7202,1007.1672,504.3458,49811.3385,0.2781,0.0015,0.0010,0.7297
std,0.5836,0.2882,1441.5325,575.0723,288.6842,28930.2895,1.1492,0.5766,0.2899,57.7872
min,-0.9998,0.0000,0.0000,0.0000,0.0000,130.0000,0.0019,-0.9996,-0.4999,-99.9000
25%,-0.5032,0.2514,1236.0000,510.0000,253.0000,24716.5000,0.0491,-0.4960,-0.2484,-48.7000
50%,-0.0062,0.5060,2496.0000,1018.0000,503.0000,49674.0000,0.0806,0.0019,0.0028,0.9000
75%,0.5135,0.7562,3723.2500,1501.0000,755.0000,74815.0000,0.1631,0.5017,0.2507,50.1000
max,0.9999,0.9999,5000.0000,2000.0000,1000.0000,99997.0000,32.2117,0.9994,0.4999,99.9000


 Analysis 1 : I looked at the raw numbers and noticed that likes_count, shares_count, and comments_count look completely linear. In real life, social media doesn't work this way—some posts get 0 likes and a few get 10k+.

The Engineering Culprit: Because this dataset is synthetic, the generator used an artificial uniform distribution. This created a massive structural flaw where a post could randomly get 5,000 likes but only 130 impressions. When calculating Engagement Rate = (Likes + Comments + Shares ) / Impressions , dividing by such a tiny number of impressions creates an impossible outlier of 3,221% (max = 32.21).

 Analysis 2: The mentions column contains 3,941 null (NaN) records, while all other columns are entirely populated.

Engineering Insight: This missingness is informative rather than corrupted. A null value implies a real-world user action: the post simply did not mention/tag another user.


##IMPORTANT NOTE

Observation: The summary statistics capture explicit negative values for metrics like sentiment_score, user_past_sentiment_avg, user_engagement_growth and buzz_change_rate.

Engineering Insight: * The sentiment columns utilize a native Polarity Scale (bounded between [-1.0, 1.0]), where -1.0 signifies maximum toxicity/negativity, 0.0 represents neutral prose, and 1.0 tracks highly positive sentiment.
The buzz_change_rate operates as a velocity vector (bounded between [-100%, 100%]). A negative value indicates a decay in content momentum and topical conversation over the tracked window.